native/src/widget/grid.rs

```diff
@@ -0,0 +1,276 @@
+//! Write some text for your users to read.
+use crate::{
+    layout::{
+        flex::{self, Axis},
+        Limits, Node,
+    },
+    Align, Element, Hasher, Layout, Length, Point, Size, Widget,
+};
+use std::{any::TypeId, hash::Hash, iter};
+
+/// A container that distributes its contents in a grid.
+///
+/// # Example
+///
+/// ```
+/// # use iced_native::{renderer::Null, Element, Grid as NativeGrid, Text};
+/// # type Grid<'a> = NativeGrid<'a, (), Null>;
+/// Grid::with_children(
+///     vec![
+///         Element::from(Text::new("First row, first column")),
+///         Element::from(Text::new("First row, second column")),
+///         Element::from(Text::new("Second row, first column")),
+///         Element::from(Text::new("Second row, second column")),
+///     ],
+/// )
+/// .columns(2);
+/// ```
+#[allow(missing_debug_implementations)]
+pub struct Grid<'a, Message, Renderer> {
+    columns: Option<usize>,
+    column_width: Option<u16>,
+    elements: Vec<Element<'a, Message, Renderer>>,
+}
+
+impl<'a, Message, Renderer> Grid<'a, Message, Renderer> {
+    /// Create a new empty [`Grid`].
+    ///
+    /// [`Grid`]: struct.Grid.html
+    pub fn new() -> Self {
+        Self::with_children(Vec::new())
+    }
+
+    /// Create a new [`Grid`] with the given [`Element`]s.
+    ///
+    /// [`Grid`]: struct.Grid.html
+    /// [`Element`]: ../struct.Element.html
+    pub fn with_children(
+        elements: Vec<Element<'a, Message, Renderer>>,
+    ) -> Self {
+        Self {
+            columns: None,
+            column_width: None,
+            elements,
+        }
+    }
+
+    /// Sets a fixed amount of columns for the [`Grid`].
+    ///
+    /// [`Grid`]: struct.Grid.html
+    pub fn columns(mut self, columns: usize) -> Self {
+        if columns == 0 {
+            self.columns = None;
+        } else {
+            self.columns = Some(columns);
+        }
+
+        self
+    }
+
+    /// Sets the width of columns for the [`Grid`].
+    ///
+    /// [`Grid`]: struct.Grid.html
+    pub fn column_width(mut self, column_width: u16) -> Self {
+        self.column_width = Some(column_width);
+        self
+    }
+
+    /// Adds an [`Element`] to the [`Grid`].
+    ///
+    /// [`Element`]: ../struct.Element.html
+    /// [`Grid`]: struct.Grid.html
+    pub fn push<E>(mut self, element: E) -> Self
+    where
+        E: Into<Element<'a, Message, Renderer>>,
+    {
+        self.elements.push(element.into());
+        self
+    }
+}
+
+impl<'a, Message, Renderer> Widget<Message, Renderer>
+    for Grid<'a, Message, Renderer>
+where
+    Renderer: self::Renderer,
+{
+    fn width(&self) -> Length {
+        if self.columns.is_some() {
+            Length::Shrink
+        } else {
+            Length::Fill
+        }
```

Comment:
I hope I got that right?

Reply:
I think it should be the other way around. When we have a fixed column width, chances are we will not be able to fill all the available space.

Reply:
Alright.